File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func mergeKLists(_ lists: [ListNode?]) -> ListNode? { //_means no need to put internal parameter name
        if(lists.isEmpty || lists.count == 0){
            return nil
        }
        var mergedList : ListNode?
        for var list in lists{
            mergedList = mergeHelper(&mergedList, &list)
        }
        return mergedList
    }

    func mergeHelper(_ mergedList: inout ListNode?, _ list: inout ListNode?) -> ListNode? {
        var starter = ListNode(0) // Create a dummy node with value 0
        var current = starter      // Pointer to build the merged list

        // Iterate while neither of the lists is exhausted
        while let n1 = mergedList, let n2 = list {
            if n1.val < n2.val {
                current.next = n1
                mergedList = n1.next // Move to next node in mergedList
            } else {
                current.next = n2
                list = n2.next // Move to next node in list
            }
            current = current.next! // Move the current pointer
        }

        // At this point, one of the lists is exhausted, we append the remainder of the other list
        current.next = mergedList ?? list
        
        return starter.next // Return the merged list, excluding the dummy node
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists == null || lists.length ==0){
            return null;
        }
        ListNode mergedList = null;
        for(ListNode list : lists){
            mergedList = mergeHelper(mergedList, list);
        }
        return mergedList; 
    }
    private ListNode mergeHelper(ListNode mergedList, ListNode list){
        ListNode head = new ListNode(0);// initialise the 1st pointer (value does not matter)
        ListNode current = head;
        //merge lists in sorted order
        while(mergedList!=null &&list!=null){ //traversing the linkedlist
            if(mergedList.val < list.val){
                current.next = mergedList;
                mergedList = mergedList.next;
            }
            else{
                current.next = list;
                list = list.next;
            }
            current = current.next; //point to next node
        }

        if(mergedList != null){
            current.next = mergedList;
        }
        else{
            current.next = list; //attaching the list to current
        }
        return head.next;//return the head of the LinkList
    }

}